prototypes/emdpg_waveguide/EM_waveguide.cc

```diff
@@ -0,0 +1,2451 @@
+#include <deal.II/base/tensor.h>
+#include <deal.II/base/timer.h>
+#include <deal.II/base/utilities.h>
+
+#include <deal.II/distributed/grid_refinement.h>
+#include <deal.II/distributed/tria.h>
+
+#include <deal.II/dofs/dof_handler.h>
+#include <deal.II/dofs/dof_tools.h>
+
+#include <deal.II/fe/fe_dgq.h>
+#include <deal.II/fe/fe_nedelec_sz.h>
+#include <deal.II/fe/fe_system.h>
+#include <deal.II/fe/fe_values.h>
+
+#include <deal.II/grid/grid_generator.h>
+#include <deal.II/grid/grid_refinement.h>
+#include <deal.II/grid/grid_tools.h>
+#include <deal.II/grid/tria.h>
+
+#include <deal.II/lac/affine_constraints.h>
+#include <deal.II/lac/dynamic_sparsity_pattern.h>
+#include <deal.II/lac/generic_linear_algebra.h>
+#include <deal.II/lac/lapack_full_matrix.h>
+#include <deal.II/lac/precondition.h>
+#include <deal.II/lac/solver_cg.h>
+#include <deal.II/lac/sparse_direct.h>
+#include <deal.II/lac/sparse_matrix.h>
+#include <deal.II/lac/sparsity_tools.h>
+#include <deal.II/lac/trilinos_precondition.h>
+#include <deal.II/lac/vector.h>
+
+#include <deal.II/numerics/data_out.h>
+#include <deal.II/numerics/data_out_faces.h>
+#include <deal.II/numerics/error_estimator.h>
+#include <deal.II/numerics/matrix_tools.h>
+#include <deal.II/numerics/vector_tools.h>
+
+#include <complex>
+#include <fstream>
+#include <iostream>
+
+/**
+ * @namespace LA
+ * @brief Alias namespace for dealii::LinearAlgebraTrilinos.
+ *
+ * This namespace provides a convenient shorthand for accessing the
+ * Trilinos-based linear algebra functionality from the deal.II library.
+ * All classes and functions from dealii::LinearAlgebraTrilinos can be
+ * accessed via the LA namespace.
+ */
+namespace LA
+{
+  using namespace dealii::LinearAlgebraTrilinos;
+} // namespace LA
+namespace EM_DPG
+{
+  using namespace dealii;
+
+  // This is a helper function that maps a 3D tensor to a 2D face in
+  // H^-1/2(curl).
+  template <int dim>
+  DEAL_II_ALWAYS_INLINE inline Tensor<1, dim, std::complex<double>>
+  map_H12(const Tensor<1, dim, std::complex<double>> &tensor,
+          const Tensor<1, dim>                       &normal)
+  {
+    auto result = cross_product_3d(normal, cross_product_3d(tensor, normal));
+
+    return result;
+  }
+
+  // We create a class to handle the input parameters.
+  template <int dim>
+  class Parameters : public ParameterAcceptor
+  {
+  public:
+    // CONSTRUCTOR
+    Parameters();
+
+    // Constexpr values
+    static constexpr double               speed_of_light = 299792458.;
+    static constexpr std::complex<double> imag{0., 1.};
+
+    // Material properties
+    std::complex<double> epsilon_r;
+    std::complex<double> mu_r;
+    double               sigma_r;
+
+    // Waveguide geometry
+    double waveguide_a; // Width of the waveguide
+    double waveguide_b; // Height of the waveguide
+    double waveguide_length;
+
+    // Frequency and excitation modes
+    double frequency;
+    int    mode_x;
+    int    mode_y;
```

Comment:
Very picky but couldn't this also be unsigned int? is there such a thing as a mode -N?
```suggestion
    unsigned int    mode_x;
    unsigned int    mode_y;
```